File: paleo_workbench/ui/pages/data_toolbar.py
from __future__ import annotations

from pathlib import Path

from PySide6.QtCore import QTimer, Signal
from PySide6.QtGui import QAction, QActionGroup, QIcon
from PySide6.QtWidgets import QHBoxLayout, QLabel, QLineEdit, QMenu, QPushButton, QWidget

from paleo_workbench.ui import tokens

_ICONS_DIR = Path(__file__).parent.parent.parent / "ui" / "assets" / "icons" / "map"


def _icon(name: str) -> QIcon:
    path = _ICONS_DIR / f"{name}.svg"
    return QIcon(str(path)) if path.exists() else QIcon()


class DataToolbar(QWidget):
    import_files_requested = Signal()
    import_folder_requested = Signal()
    rescan_requested = Signal()
    remove_requested = Signal()
    open_folder_requested = Signal()
    visualize_requested = Signal()
    clear_preview_cache_requested = Signal()
    verify_requested = Signal()
    # Data health entry: open the Catalog Health (audit) dialog.
    health_check_requested = Signal()
    reader_toggled = Signal()
    search_changed = Signal(str)
    # Multi-tag asset-table filter: (selected tag names, "and"|"or").
    tag_filter_changed = Signal(list, str)
    # Open the Tag Manager dialog (tags CRUD / merge / prune).
    tag_manager_requested = Signal()

    def __init__(self, parent=None):
        super().__init__(parent)
        self.setObjectName("DataToolbar")

        layout = QHBoxLayout(self)
        layout.setContentsMargins(0, 0, 0, 0)
        layout.setSpacing(tokens.SPACE_1)

        self.import_btn = QPushButton(_icon("btn-import"), "导入文件")
        self.import_btn.setObjectName("PrimaryButton")
        self.import_btn.setMinimumHeight(tokens.CONTROL_HEIGHT_LG)
        self.import_btn.setToolTip("导入文件并创建项目受管的不可变 RAW 副本")
        self.import_btn.clicked.connect(self.import_files_requested.emit)
        layout.addWidget(self.import_btn)

        self.import_folder_btn = QPushButton(_icon("btn-import-folder"), "导入目录")
        self.import_folder_btn.setObjectName("SecondaryButton")
        self.import_folder_btn.setMinimumHeight(tokens.CONTROL_HEIGHT)
        self.import_folder_btn.setToolTip("导入整个目录")
        self.import_folder_btn.clicked.connect(self.import_folder_requested.emit)
        layout.addWidget(self.import_folder_btn)

        self.verify_btn = QPushButton(_icon("btn-verify"), "完整性校验")
        self.verify_btn.setObjectName("SecondaryButton")
        self.verify_btn.setMinimumHeight(tokens.CONTROL_HEIGHT)
        self.verify_btn.setToolTip("后台校验数据资产完整性与 SHA-256")
        self.verify_btn.clicked.connect(self.verify_requested.emit)
        layout.addWidget(self.verify_btn)
        self._verify_running = False

        self.health_btn = QPushButton(_icon("btn-health"), "健康检查")
        self.health_btn.setObjectName("SecondaryButton")
        self.health_btn.setMinimumHeight(tokens.CONTROL_HEIGHT)
        self.health_btn.setToolTip("数据目录健康体检：资产/版本统计、缺失、血缘断链、标签悬挂、孤儿文件")
        self.health_btn.clicked.connect(self.health_check_requested.emit)
        layout.addWidget(self.health_btn)

        self.rescan_btn = QPushButton(_icon("btn-rescan"), "重新扫描")
        self.rescan_btn.setObjectName("SecondaryButton")
        self.rescan_btn.setMinimumHeight(tokens.CONTROL_HEIGHT)
        self.rescan_btn.setToolTip("重新扫描选中项")
        self.rescan_btn.clicked.connect(self.rescan_requested.emit)
        layout.addWidget(self.rescan_btn)

        self.remove_btn = QPushButton(_icon("btn-remove"), "移出项目")
        self.remove_btn.setObjectName("SecondaryButton")
        self.remove_btn.setMinimumHeight(tokens.CONTROL_HEIGHT)
        self.remove_btn.setToolTip("移出项目（不删源文件）")
        self.remove_btn.clicked.connect(self.remove_requested.emit)
        layout.addWidget(self.remove_btn)

        self.open_folder_btn = QPushButton(_icon("btn-open-folder"), "打开目录")
        self.open_folder_btn.setObjectName("SecondaryButton")
        self.open_folder_btn.setMinimumHeight(tokens.CONTROL_HEIGHT)
        self.open_folder_btn.setToolTip("在文件管理器中打开")
        self.open_folder_btn.clicked.connect(self.open_folder_requested.emit)
        layout.addWidget(self.open_folder_btn)

        self.visualize_btn = QPushButton(_icon("btn-visualize"), "可视化")
        self.visualize_btn.setObjectName("SecondaryButton")
        self.visualize_btn.setMinimumHeight(tokens.CONTROL_HEIGHT)
        self.visualize_btn.setToolTip("在可视化页面打开")
        self.visualize_btn.clicked.connect(self.visualize_requested.emit)
        layout.addWidget(self.visualize_btn)

        self.clear_preview_cache_btn = QPushButton(_icon("btn-clear-cache"), "清除预览缓存")
        self.clear_preview_cache_btn.setObjectName("SecondaryButton")
        self.clear_preview_cache_btn.setMinimumHeight(tokens.CONTROL_HEIGHT)
        self.clear_preview_cache_btn.setToolTip("清除项目预览磁盘缓存")
        self.clear_preview_cache_btn.clicked.connect(
            self.clear_preview_cache_requested.emit
        )
        layout.addWidget(self.clear_preview_cache_btn)

        # --- Tag tools -------------------------------------------------------
        # 标签筛选: checkable tag list + AND/OR + clear, applied immediately.
        self.tag_filter_btn = QPushButton(_icon("btn-tag-filter"), "标签筛选")
        self.tag_filter_btn.setObjectName("SecondaryButton")
        self.tag_filter_btn.setMinimumHeight(tokens.CONTROL_HEIGHT)
        self.tag_filter_btn.setToolTip("按标签筛选资产表（支持多选与 AND/OR 组合）")
        self._tag_filter_menu = QMenu(self.tag_filter_btn)
        self._tag_filter_menu.aboutToShow.connect(self._rebuild_tag_filter_menu)
        self.tag_filter_btn.setMenu(self._tag_filter_menu)
        layout.addWidget(self.tag_filter_btn)

        self.tag_manager_btn = QPushButton(_icon("btn-tag-manager"), "标签管理")
        self.tag_manager_btn.setObjectName("SecondaryButton")
        self.tag_manager_btn.setMinimumHeight(tokens.CONTROL_HEIGHT)
        self.tag_manager_btn.setToolTip("管理标签：新建 / 重命名 / 合并 / 清理")
        self.tag_manager_btn.clicked.connect(self.tag_manager_requested.emit)
        layout.addWidget(self.tag_manager_btn)

        self.operation_status_label = QLabel("")
        self.operation_status_label.setStyleSheet(f"color: {tokens.TEXT_SECONDARY};")
        layout.addWidget(self.operation_status_label)

        # Tag filter panel state (menu rebuilt lazily from candidates).
        self._tag_candidates: list[str] = []
        self._selected_tags: list[str] = []
        self._tag_operator: str = "and"
        self._tag_check_actions: list[QAction] = []

        self._search_timer = QTimer(self)
        self._search_timer.setSingleShot(True)
        self._search_timer.setInterval(180)
        self._pending_search = ""
        self._search_timer.timeout.connect(self._emit_debounced_search)

        self.search_box = QLineEdit()
        self.search_box.setObjectName("SearchBox")
        self.search_box.setPlaceholderText("搜索文件名 / 类型 / 阶段 / 标签 / 路径...")
        self.search_box.setToolTip("搜索文件名/类型/阶段/标签/路径")
        self.search_box.setClearButtonEnabled(True)
        self.search_box.setMinimumHeight(tokens.CONTROL_HEIGHT)
        self.search_box.textChanged.connect(self._on_search_text_changed)
        layout.addWidget(self.search_box, 1)

        self.column_settings_slot = QWidget(self)
        self.column_settings_slot.setObjectName("ColumnSettingsSlot")
        column_settings_layout = QHBoxLayout(self.column_settings_slot)
        column_settings_layout.setContentsMargins(0, 0, 0, 0)
        column_settings_layout.setSpacing(0)
        layout.addWidget(self.column_settings_slot)

        # Hides the whole right column (reader + inspector), not only the reader pane.
        self.reader_btn = QPushButton(_icon("btn-reader"), "预览栏")
        self.reader_btn.setObjectName("SecondaryButton")
        self.reader_btn.setMinimumHeight(tokens.CONTROL_HEIGHT)
        self.reader_btn.setCheckable(True)
        self.reader_btn.setToolTip("显示或隐藏右侧预览与属性面板")
        self.reader_btn.clicked.connect(self.reader_toggled.emit)
        layout.addWidget(self.reader_btn)

    def set_verify_running(self, running: bool) -> None:
        self._verify_running = bool(running)
        if self._verify_running:
            self.verify_btn.setText("取消校验")
            self.verify_btn.setToolTip("取消正在进行的完整性校验")
        else:
            self.verify_btn.setText("完整性校验")
            self.verify_btn.setToolTip("后台校验数据资产完整性与 SHA-256")

    def set_column_settings_button(self, button: QPushButton) -> None:
        self.column_settings_slot.layout().addWidget(button)

    # --- Tag filter panel -----------------------------------------------------

    def set_tag_candidates(self, tags: list[str]) -> None:
        """Provide the selectable tag list (from FilterIndex counts or the
        catalog service). Rebuilds the menu on next open."""
        self._tag_candidates = sorted(
            {t for t in tags if t and str(t).strip()}
        )

    def tag_candidates(self) -> list[str]:
        return list(self._tag_candidates)

    def set_tag_operator(self, operator: str) -> None:
        if operator in ("and", "or"):
            self._tag_operator = operator

    def current_tag_selection(self) -> list[str]:
        # Read the filter state, not the lazily-built menu-action mirror: the
        # actions only exist after the menu was opened once, so reading them
        # here would silently report [] for programmatic applies (Tag Manager
        # 查看关联数据) and drop the tag filter on the next navigation click
        # (#413). The menu rebuild initializes its check states from
        # _selected_tags, so the mirror can never diverge from the source.
        return list(self._selected_tags)

    def current_tag_operator(self) -> str:
        return self._tag_operator

    def apply_tag_selection(self, tags: list[str], operator: str = "and") -> None:
        """Programmatically set the checked tag filter (single source of truth
        for the visible filter state, e.g. Tag Manager 双击查看关联数据)."""
        self._selected_tags = [t for t in tags if t and str(t).strip()]
        if operator in ("and", "or"):
            self._tag_operator = operator
        self._emit_tag_filter()

    def _rebuild_tag_filter_menu(self) -> None:
        self._tag_filter_menu.clear()
        self._tag_check_actions = []

        if not self._tag_candidates:
            empty = QAction("暂无可用标签", self._tag_filter_menu)
            empty.setEnabled(False)
            self._tag_filter_menu.addAction(empty)
            return

        for tag in self._tag_candidates:
            action = QAction(tag, self._tag_filter_menu)
            action.setCheckable(True)
            action.setChecked(tag in self._selected_tags)
            action.toggled.connect(
                lambda checked, name=tag: self._on_tag_toggled(name, checked)
            )
            self._tag_filter_menu.addAction(action)
            self._tag_check_actions.append(action)

        self._tag_filter_menu.addSeparator()

        operator_group = QActionGroup(self._tag_filter_menu)
        operator_group.setExclusive(True)
        for label, value in (("匹配全部 (AND)", "and"), ("匹配任一 (OR)", "or")):
            op_action = QAction(label, self._tag_filter_menu)
            op_action.setCheckable(True)
            op_action.setChecked(self._tag_operator == value)
            op_action.setActionGroup(operator_group)
            op_action.toggled.connect(
                lambda checked, op=value: self._on_operator_changed(op, checked)
            )
            self._tag_filter_menu.addAction(op_action)

        self._tag_filter_menu.addSeparator()

        clear_action = QAction("清除标签筛选", self._tag_filter_menu)
        clear_action.triggered.connect(self.clear_tag_filter)
        self._tag_filter_menu.addAction(clear_action)

    def _on_tag_toggled(self, tag_name: str, checked: bool) -> None:
        if checked and tag_name not in self._selected_tags:
            self._selected_tags.append(tag_name)
        elif not checked and tag_name in self._selected_tags:
            self._selected_tags.remove(tag_name)
        self._emit_tag_filter()

    def _on_operator_changed(self, operator: str, checked: bool) -> None:
        if not checked:
            return
        self._tag_operator = operator
        self._emit_tag_filter()

    def clear_tag_filter(self) -> None:
        self._selected_tags = []
        self._emit_tag_filter()

    def _emit_tag_filter(self) -> None:
        self.tag_filter_changed.emit(list(self._selected_tags), self._tag_operator)

    def _on_search_text_changed(self, text: str) -> None:
        self._pending_search = text
        self._search_timer.start()

    def _emit_debounced_search(self) -> None:
        self.search_changed.emit(self._pending_search)

File: libs/ui_pages_data/src/qt/data_toolbar.cpp
// UI-06 — DataToolbar shell (see qt/data_toolbar.hpp).
#include <pwb/ui_pages_data/qt/data_toolbar.hpp>

#include <QAction>
#include <QActionGroup>
#include <QComboBox>
#include <QHBoxLayout>
#include <QLabel>
#include <QLineEdit>
#include <QMenu>
#include <QPushButton>
#include <QTimer>
#include <QVBoxLayout>

#include <algorithm>

#include <pwb/ui_pages_data/timings.hpp>
#include <pwb/ui_shell/style_registry.hpp>

namespace pwb::ui_pages_data::qt {

ToolbarIconFn DataToolbar::icon_provider_;

void DataToolbar::set_icon_provider(ToolbarIconFn fn) {
    icon_provider_ = std::move(fn);
}

namespace {

QIcon icon(const std::string& name) {
    return DataToolbar::icon_provider() ? DataToolbar::icon_provider()(name)
                                        : QIcon();
}

// sorted({t for t in tags if t and str(t).strip()})
std::vector<std::string> clean_tags(const std::vector<std::string>& tags) {
    std::vector<std::string> out;
    for (const auto& tag : tags) {
        const auto first = tag.find_first_not_of(" \t\r\n");
        if (tag.empty() || first == std::string::npos) continue;
        if (std::find(out.begin(), out.end(), tag) == out.end()) {
            out.push_back(tag);
        }
    }
    std::sort(out.begin(), out.end());
    return out;
}

QPushButton* make_button(QWidget* parent, const char* icon_name,
                         const QString& text, const QString& object,
                         const QString& tooltip) {
    auto* btn = new QPushButton(icon(icon_name), text, parent);
    btn->setObjectName(object);
    ui_shell::style_track_control_height(btn);
    btn->setToolTip(tooltip);
    return btn;
}

}  // namespace

DataToolbar::DataToolbar(QWidget* parent) : QWidget(parent) {
    setObjectName(QStringLiteral("DataToolbar"));
    auto* layout = new QVBoxLayout(this);
    layout->setContentsMargins(0, 0, 0, 0);
    layout->setSpacing(4);                       // SPACE_1

    // 稿式中列头：「数据列表」标题 + 过滤行。页面级功能按钮已由
    // Ribbon 命令承载（emit 同名信号），不再占行——全部创建后隐藏，
    // 仅作信号源保留（命令面板/快捷键复用同一信号身份）。
    auto* title_row = new QHBoxLayout();
    title_row->setContentsMargins(0, 0, 0, 0);
    title_row->setSpacing(4);
    auto* title = new QLabel(QStringLiteral("数据列表"), this);
    title->setObjectName(QStringLiteral("DataListTitle"));
    {
        const auto p = ui_shell::style_palette();
        const auto it = p.find("TEXT_PRIMARY");
        title->setStyleSheet(QStringLiteral(
            "font-weight: 600; color: %1;")
                .arg(it != p.end() ? QString::fromStdString(it->second)
                                   : QString()));
    }
    title_row->addWidget(title);
    title_row->addStretch(1);
    column_settings_slot_ = new QWidget(this);
    column_settings_slot_->setObjectName(QStringLiteral("ColumnSettingsSlot"));
    auto* slot_layout = new QHBoxLayout(column_settings_slot_);
    slot_layout->setContentsMargins(0, 0, 0, 0);
    slot_layout->setSpacing(0);
    title_row->addWidget(column_settings_slot_);
    layout->addLayout(title_row);

    auto* filter_row = new QHBoxLayout();
    filter_row->setContentsMargins(0, 0, 0, 0);
    filter_row->setSpacing(4);

    import_btn_ = new QPushButton(icon("btn-import"),
                                  QStringLiteral("导入文件"), this);
    import_btn_->setObjectName(QStringLiteral("PrimaryButton"));
    import_btn_->setMinimumHeight(34);           // CONTROL_HEIGHT_LG
    import_btn_->setToolTip(
        QStringLiteral("导入文件并创建项目受管的不可变 RAW 副本"));
    connect(import_btn_, &QPushButton::clicked, this,
            &DataToolbar::import_files_requested);
    import_btn_->setVisible(false);

    import_folder_btn_ = make_button(
        this, "btn-import-folder", QStringLiteral("导入目录"),
        "SecondaryButton", QStringLiteral("导入整个目录"));
    connect(import_folder_btn_, &QPushButton::clicked, this,
            &DataToolbar::import_folder_requested);
    import_folder_btn_->setVisible(false);  // 稿式行不承载——Ribbon 同名命令接此信号

    plan_import_btn_ = make_button(
        this, "btn-import", QStringLiteral("规划导入…"), "SecondaryButton",
        QStringLiteral(
            "规划导入：扫描/分类/身份匹配/查重 → 逐项确认 → 分块执行"
            "（与 Agent 导入共用同一计划服务）"));
    connect(plan_import_btn_, &QPushButton::clicked, this,
            &DataToolbar::plan_import_requested);
    plan_import_btn_->setVisible(false);  // 稿式行不承载——Ribbon 同名命令接此信号

    // D4: visible only while an import runs; cooperative cancel.
    cancel_import_btn_ = new QPushButton(QStringLiteral("取消导入"), this);
    cancel_import_btn_->setObjectName(QStringLiteral("SecondaryButton"));
    cancel_import_btn_->setMinimumHeight(30);    // CONTROL_HEIGHT
    cancel_import_btn_->setToolTip(QStringLiteral(
        "协作式取消：当前分块完成后停止，已完成分块保持一致"));
    cancel_import_btn_->setVisible(false);
    connect(cancel_import_btn_, &QPushButton::clicked, this,
            &DataToolbar::cancel_import_requested);
    cancel_import_btn_->setVisible(false);

    verify_btn_ = make_button(this, "btn-verify", QStringLiteral("完整性校验"),
                              "SecondaryButton",
                              QStringLiteral("后台校验数据资产完整性与 SHA-256"));
    connect(verify_btn_, &QPushButton::clicked, this,
            &DataToolbar::verify_requested);
    verify_btn_->setVisible(false);  // 稿式行不承载——Ribbon 同名命令接此信号

    health_btn_ = make_button(
        this, "btn-health", QStringLiteral("健康检查"), "SecondaryButton",
        QStringLiteral(
            "数据目录健康体检：资产/版本统计、缺失、血缘断链、标签悬挂、孤儿文件"));
    connect(health_btn_, &QPushButton::clicked, this,
            &DataToolbar::health_check_requested);
    health_btn_->setVisible(false);  // 稿式行不承载——Ribbon 同名命令接此信号

    rescan_btn_ = make_button(this, "btn-rescan", QStringLiteral("重新扫描"),
                              "SecondaryButton", QStringLiteral("重新扫描选中项"));
    connect(rescan_btn_, &QPushButton::clicked, this,
            &DataToolbar::rescan_requested);
    rescan_btn_->setVisible(false);  // 稿式行不承载——Ribbon 同名命令接此信号

    remove_btn_ = make_button(this, "btn-remove", QStringLiteral("移出项目"),
                              "SecondaryButton",
                              QStringLiteral("移出项目（不删源文件）"));
    connect(remove_btn_, &QPushButton::clicked, this,
            &DataToolbar::remove_requested);
    remove_btn_->setVisible(false);  // 稿式行不承载——Ribbon 同名命令接此信号

    open_folder_btn_ = make_button(this, "btn-open-folder",
                                   QStringLiteral("打开目录"), "SecondaryButton",
                                   QStringLiteral("在文件管理器中打开"));
    connect(open_folder_btn_, &QPushButton::clicked, this,
            &DataToolbar::open_folder_requested);
    open_folder_btn_->setVisible(false);  // 稿式行不承载——Ribbon 同名命令接此信号

    visualize_btn_ = make_button(this, "btn-visualize", QStringLiteral("可视化"),
                                 "SecondaryButton",
                                 QStringLiteral("在可视化页面打开"));
    connect(visualize_btn_, &QPushButton::clicked, this,
            &DataToolbar::visualize_requested);
    visualize_btn_->setVisible(false);  // 稿式行不承载——Ribbon 同名命令接此信号

    clear_preview_cache_btn_ = make_button(
        this, "btn-clear-cache", QStringLiteral("清除预览缓存"),
        "SecondaryButton", QStringLiteral("清除项目预览磁盘缓存"));
    connect(clear_preview_cache_btn_, &QPushButton::clicked, this,
            &DataToolbar::clear_preview_cache_requested);
    clear_preview_cache_btn_->setVisible(false);  // 稿式行不承载——Ribbon 同名命令接此信号

    // --- Tag tools ---------------------------------------------------------
    tag_filter_btn_ = make_button(this, "btn-tag-filter",
                                  QStringLiteral("标签筛选"), "SecondaryButton",
                                  QStringLiteral("按标签筛选资产表（支持多选与 AND/OR 组合）"));
    tag_filter_menu_ = new QMenu(tag_filter_btn_);
    connect(tag_filter_menu_, &QMenu::aboutToShow, this,
            &DataToolbar::rebuild_tag_filter_menu);
    tag_filter_btn_->setMenu(tag_filter_menu_);
    tag_filter_btn_->setVisible(false);  // 稿式行不承载——Ribbon 同名命令接此信号

    tag_manager_btn_ = make_button(this, "btn-tag-manager",
                                   QStringLiteral("标签管理"), "SecondaryButton",
                                   QStringLiteral("管理标签：新建 / 重命名 / 合并 / 清理"));
    connect(tag_manager_btn_, &QPushButton::clicked, this,
            &DataToolbar::tag_manager_requested);
    tag_manager_btn_->setVisible(false);  // 稿式行不承载——Ribbon 同名命令接此信号

    operation_status_label_ = new QLabel(QString(), this);
    {
        const auto p = ui_shell::style_palette();
        const auto it = p.find("TEXT_SECONDARY");
        if (it != p.end()) {
            operation_status_label_->setStyleSheet(
                QStringLiteral("color: %1;")
                    .arg(QString::fromStdString(it->second)));
        }
    }
    operation_status_label_->setVisible(false);

    search_timer_ = new QTimer(this);
    search_timer_->setSingleShot(true);
    search_timer_->setInterval(kSearchDebounceMs);  // 180
    connect(search_timer_, &QTimer::timeout, this,
            &DataToolbar::emit_debounced_search);

    search_box_ = new QLineEdit(this);
    search_box_->setObjectName(QStringLiteral("SearchBox"));
    search_box_->setPlaceholderText(
        QStringLiteral("搜索名称、井名、来源…"));
    search_box_->setToolTip(
        QStringLiteral("搜索名称、井名、来源"));
    search_box_->setClearButtonEnabled(true);
    ui_shell::style_track_control_height(search_box_);
    connect(search_box_, &QLineEdit::textChanged, this,
            [this](const QString& text) {
                if (search_sync_) return;
                pending_search_ = text;
                search_timer_->start();
            });
    filter_row->addWidget(search_box_, 1);

    // 稿：所有类型 ▼ / 所有状态 ▼ —— 候选由真实行集驱动
    // （set_filter_options），空工程只剩「所有*」。
    type_combo_ = new QComboBox(this);
    type_combo_->setObjectName(QStringLiteral("DataTypeFilter"));
    type_combo_->addItem(QStringLiteral("所有类型"), QString());
    connect(type_combo_, &QComboBox::currentIndexChanged, this,
            [this](int index) {
                Q_EMIT type_filter_changed(
                    type_combo_->itemData(index).toString());
            });
    filter_row->addWidget(type_combo_);

    status_combo_ = new QComboBox(this);
    status_combo_->setObjectName(QStringLiteral("DataStatusFilter"));
    status_combo_->addItem(QStringLiteral("所有状态"), QString());
    connect(status_combo_, &QComboBox::currentIndexChanged, this,
            [this](int index) {
                Q_EMIT status_filter_changed(
                    status_combo_->itemData(index).toString());
            });
    filter_row->addWidget(status_combo_);

    filter_row->addStretch(1);

    // 稿：行尾计数「共 N 条数据」。
    count_label_ = new QLabel(QStringLiteral("共 0 条数据"), this);
    count_label_->setObjectName(QStringLiteral("DataRowCount"));
    {
        const auto p = ui_shell::style_palette();
        const auto it = p.find("TEXT_SECONDARY");
        if (it != p.end()) {
            count_label_->setStyleSheet(
                QStringLiteral("color: %1;")
                    .arg(QString::fromStdString(it->second)));
        }
    }
    filter_row->addWidget(count_label_);
    layout->addLayout(filter_row);

    // Hides the whole right column (reader + inspector), not only the reader.
    reader_btn_ = make_button(this, "btn-reader", QStringLiteral("预览栏"),
                              "SecondaryButton",
                              QStringLiteral("显示或隐藏右侧预览与属性面板"));
    reader_btn_->setCheckable(true);
    connect(reader_btn_, &QPushButton::clicked, this,
            &DataToolbar::reader_toggled);
    reader_btn_->setVisible(false);
}

void DataToolbar::set_verify_running(bool running) {
    verify_running_ = running;
    const VerifyButtonState state = verify_button_state(running);
    verify_btn_->setText(QString::fromStdString(state.text));
    verify_btn_->setToolTip(QString::fromStdString(state.tooltip));
}

void DataToolbar::set_column_settings_button(QPushButton* button) {
    column_settings_slot_->layout()->addWidget(button);
}

// --- tag filter --------------------------------------------------------------

void DataToolbar::set_tag_candidates(const std::vector<std::string>& tags) {
    tag_candidates_ = clean_tags(tags);
}

void DataToolbar::set_tag_operator(const std::string& op) {
    if (op == "and" || op == "or") tag_operator_ = op;
}

void DataToolbar::apply_tag_selection(const std::vector<std::string>& tags,
                                      const std::string& op) {
    selected_tags_.clear();
    for (const auto& tag : tags) {
        const auto first = tag.find_first_not_of(" \t\r\n");
        if (!tag.empty() && first != std::string::npos) {
            selected_tags_.push_back(tag);
        }
    }
    if (op == "and" || op == "or") tag_operator_ = op;
    emit_tag_filter();
}

void DataToolbar::rebuild_tag_filter_menu() {
    tag_filter_menu_->clear();
    tag_check_actions_.clear();
    if (tag_candidates_.empty()) {
        auto* empty = new QAction(QStringLiteral("暂无可用标签"),
                                  tag_filter_menu_);
        empty->setEnabled(false);
        tag_filter_menu_->addAction(empty);
        return;
    }
    for (const auto& tag : tag_candidates_) {
        auto* action =
            new QAction(QString::fromStdString(tag), tag_filter_menu_);
        action->setCheckable(true);
        action->setChecked(std::find(selected_tags_.begin(),
                                     selected_tags_.end(),
                                     tag) != selected_tags_.end());
        connect(action, &QAction::toggled, this,
                [this, tag](bool checked) { on_tag_toggled(tag, checked); });
        tag_filter_menu_->addAction(action);
        tag_check_actions_.push_back(action);
    }
    tag_filter_menu_->addSeparator();

    auto* operator_group = new QActionGroup(tag_filter_menu_);
    operator_group->setExclusive(true);
    const std::pair<const char*, const char*> ops[] = {
        {"匹配全部 (AND)", "and"}, {"匹配任一 (OR)", "or"}};
    for (const auto& [label, value] : ops) {
        auto* op_action = new QAction(QString::fromUtf8(label),
                                      tag_filter_menu_);
        op_action->setCheckable(true);
        op_action->setChecked(tag_operator_ == value);
        op_action->setActionGroup(operator_group);
        const std::string op = value;
        connect(op_action, &QAction::toggled, this,
                [this, op](bool checked) { on_operator_changed(op, checked); });
        tag_filter_menu_->addAction(op_action);
    }
    tag_filter_menu_->addSeparator();

    auto* clear_action = new QAction(QStringLiteral("清除标签筛选"),
                                     tag_filter_menu_);
    connect(clear_action, &QAction::triggered, this,
            &DataToolbar::clear_tag_filter);
    tag_filter_menu_->addAction(clear_action);
}

void DataToolbar::on_tag_toggled(const std::string& tag, bool checked) {
    const auto it =
        std::find(selected_tags_.begin(), selected_tags_.end(), tag);
    if (checked && it == selected_tags_.end()) {
        selected_tags_.push_back(tag);
    } else if (!checked && it != selected_tags_.end()) {
        selected_tags_.erase(it);
    }
    emit_tag_filter();
}

void DataToolbar::on_operator_changed(const std::string& op, bool checked) {
    if (!checked) return;
    tag_operator_ = op;
    emit_tag_filter();
}

void DataToolbar::clear_tag_filter() {
    selected_tags_.clear();
    emit_tag_filter();
}

void DataToolbar::emit_tag_filter() {
    QStringList tags;
    for (const auto& tag : selected_tags_) {
        tags << QString::fromStdString(tag);
    }
    Q_EMIT tag_filter_changed(tags, QString::fromStdString(tag_operator_));
}

void DataToolbar::set_search_text_silent(const QString& text) {
    search_sync_ = true;
    search_box_->setText(text);
    search_sync_ = false;
}

void DataToolbar::set_filter_options(
    const QList<QPair<QString, QString>>& types,
    const QList<QPair<QString, QString>>& statuses) {
    // 候选 = 真实行的 distinct 值（原始键作 userData，显示名作
    // 文本）；当前选中经 userData 保留。
    const auto refill = [](QComboBox* combo, const QString& all_label,
                           const QList<QPair<QString, QString>>& values) {
        const QString current = combo->currentData().toString();
        combo->blockSignals(true);
        combo->clear();
        combo->addItem(all_label, QString());
        for (const auto& [key, label] : values) {
            combo->addItem(label, key);
        }
        const int restore = combo->findData(current);
        combo->setCurrentIndex(restore >= 0 ? restore : 0);
        combo->blockSignals(false);
    };
    refill(type_combo_, QStringLiteral("所有类型"), types);
    refill(status_combo_, QStringLiteral("所有状态"), statuses);
}

void DataToolbar::set_row_count(int count) {
    count_label_->setText(QStringLiteral("共 %1 条数据").arg(count));
}

void DataToolbar::emit_debounced_search() {
    Q_EMIT search_changed(pending_search_);
}

}  // namespace pwb::ui_pages_data::qt
